Implement distributed snapshotting

nearora-msft opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
3. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test

/kind feature

> /kind flake

**What this PR does / why we need it**:
This PR will allow snapshot sidecar controller to be deployed on every node along with the CSI drivers that handle local volumes:
1. Added a parameter "--node-deployment", which should be set to true when the sidecar is being deployed on a per node basis.
2. For these changes to work, NODE_NAME environment variable must also be set while deploying the sidecar controller.
3. With these changes, the common snapshot controller checks the node affinity for the PV related to the given VolumeSnapshot. The node affinity is matched against a list of nodes and a matching node is found. The nodeName for this matching Node is added as a label to the VolumeSnapshotContent as "snapshot.storage.kubernetes.io/managed-by=nodeName".
4. The locally deployed sidecar controller filters out the VolumeSnapshotContent objects based on the label mentioned above and will only pick those objects whose labels match with the node that the sidecar is deployed on.
5. This change also adds more rules to the RBAC settings for the common snapshot-controller so that it can get/list node related information.


**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #484 

**Special notes for your reviewer**:
Pending changes: Unit tests and documentation.

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Adds support for distributed snapshotting.
```

nearora-msft commented:
/test all

xing-yang commented:
/assign @yuxiangqian

xing-yang commented:
Does csi hostpath driver support distributed snapshotting?

nearora-msft commented:
> Does csi hostpath driver support distributed snapshotting?

I tested these changes with csi hostpath driver and it works as expected.

xing-yang commented on deploy/kubernetes/snapshot-controller/rbac-snapshot-controller.yaml:
```diff
@@ -41,7 +41,9 @@ rules:
     verbs: ["update"]
-
+  - apiGroups: [""]
+    resources: ["nodes"]
```
This new RBAC rule is only needed for distributed mode.  Can you comment it out and add comments on when it is needed?

nearora-msft replied:
Commented it out and added comments on when to use.

xing-yang commented on README.md:
```diff
@@ -165,6 +165,8 @@ Read more about how to install the example webhook [here](deploy/kubernetes/webh
 
 * `--worker-threads`: Number of worker threads for running create snapshot and delete snapshot operations. Default value is 10.
 
+* `--node-deployment`: Enables deploying the sidecar controller together with a CSI driver on nodes to manage node-local volumes. Off by default.
```
Can you add more instructions in the "Usage" section on how to use this properly.

xing-yang commented:
> I tested these changes with csi hostpath driver and it works as expected.

@nearora-msft Can you please provide more details on how you deployed it and how you tested it.  Some testing results will be helpful as well.

yuxiangqian commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -1591,3 +1602,31 @@ func (ctrl *csiSnapshotCommonController) checkAndSetInvalidSnapshotLabel(snapsho
+	nodes, err := ctrl.client.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		klog.Errorf("failed to get the list of nodes: %q", err)
+		return "", nil
```
this should return an error rather? otherwise a wrong content will be created

nearora-msft replied:
If we want to return an error here, we'll have to do that under the condition if distributed snapshotting is enabled. We'll probably have to pass a flag to this controller to identify that.

Right now, the snapshot_controller code path is same for both distributed and non distributed snapshotting and there's no distinction between these two scenarios. I thought we should keep it that way as we'll have only 1 centralized instance of this controller running even in case of distributed snapshotting. So this controller doesn't need to know if the snapshotter is being deployed on a per node basis or not. WDYT?

nearora-msft replied:
Returning an error here now.

yuxiangqian commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -1591,3 +1602,31 @@ func (ctrl *csiSnapshotCommonController) checkAndSetInvalidSnapshotLabel(snapsho
+		return "", nil
+	}
+
+	for _, node := range nodes.Items {
```
The fact that the first matching node will ALWAYS be picked to be labeled is a bit concerning though TBH I do not have a solution. It might result in a situation all snapshot contents will be labeled with the same node name, and the sidecar running on that node became hot.

nearora-msft replied:
Can there be multiple nodes that can match with the node affinity for a local volume?

awels replied:
AFAIK there can be only a single node that matches the node affinity

yuxiangqian commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -1591,3 +1602,31 @@ func (ctrl *csiSnapshotCommonController) checkAndSetInvalidSnapshotLabel(snapsho
+	}
+
+	for _, node := range nodes.Items {
+		val, _ := corev1.MatchNodeSelectorTerms(&node, nodeSelectorTerms)
```
val -> match?

yuxiangqian commented on pkg/utils/util.go:
```diff
@@ -106,7 +106,8 @@ const (
 	// See https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/177-volume-snapshot/tighten-validation-webhook-crd.md#automatic-labelling-of-invalid-objects
-	VolumeSnapshotInvalidLabel = "snapshot.storage.kubernetes.io/invalid-snapshot-resource"
+	VolumeSnapshotInvalidLabel          = "snapshot.storage.kubernetes.io/invalid-snapshot-resource"
+	VolumeSnapshotContentManagedByLabel = "csi.storage.k8s.io/managed-by"
```
use "snapshot.storage.kubernetes.io/managed_by"?

nearora-msft replied:
Done

yuxiangqian reviewed (commented):
I assume this is just part of the CL?

awels commented:
@nearora-msft Are you still working on this? If not, I would be willing to complete the work.

nearora-msft commented:
> @nearora-msft Are you still working on this? If not, I would be willing to complete the work.

Yes, sorry about the delay. Yes, still working on it. Planning to addressing the comments today.

nearora-msft commented:
> Hi @nearora-msft Can you please address the comments? Thanks.

Yes, will do

nearora-msft commented on cmd/csi-snapshotter/main.go:
```diff
@@ -116,6 +121,19 @@ func main() {
+	if *enableNodeDeployment {
+		node := os.Getenv("NODE_NAME")
+		if node == "" {
+			klog.Fatal("The NODE_NAME environment variable must be set when using --enable-node-deployment.")
```
Should there be a specific exit statement? klog.Fatal logs the error and then calls os.Exit(255).

nearora-msft commented on cmd/csi-snapshotter/main.go:
```diff
@@ -72,11 +76,12 @@ var (
+	metricsPath          = flag.String("metrics-path", "/metrics", "The HTTP path where prometheus metrics will be exposed. Default is `/metrics`.")
+	retryIntervalStart   = flag.Duration("retry-interval-start", time.Second, "Initial retry interval of failed volume snapshot creation or deletion. It doubles with each failure, up to retry-interval-max. Default is 1 second.")
+	retryIntervalMax     = flag.Duration("retry-interval-max", 5*time.Minute, "Maximum retry interval of failed volume snapshot creation or deletion. Default is 5 minutes.")
+	enableNodeDeployment = flag.Bool("node-deployment", false, "Enables deploying the sidecar controller together with a CSI driver on nodes to manage snapshots for node-local volumes.")
```
Added an exit statement below

nearora-msft commented:
> > I tested these changes with csi hostpath driver and it works as expected.
> 
> @nearora-msft Can you please provide more details on how you deployed it and how you tested it. Some testing results will be helpful as well.

I used [this](https://github.com/kubernetes-csi/csi-driver-host-path/tree/master/deploy/kubernetes-distributed) to test these changes. Basically, deployed the hostpath driver using the deploy.sh script on an Azure Kubernetes cluster and then used app-generic-ephemeral.yaml to create an app with a local volume. After that, I created a VolumeSnapshotClass and VolumeSnapshot  based on that VolumeSnapshotClass. 

As a result, the VolumeSnapshotContent CRI was handled by the instance of the snapshotter deployed on he node that had created the local volume. Also, the CreateSnapshot Grpc call was handled by the instance of hostpath driver deployed on the same node.

I hope that makes sense?

awels commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -671,6 +672,16 @@ func (ctrl *csiSnapshotCommonController) createSnapshotContent(snapshot *crdv1.V
+	if err != nil {
+		return nil, err
+	}
+	if nodeName != "" {
```
Why not simply do
```go
if node := os.Getenv("NODE_NAME"); node != "" {
  snapshotContent.Labels = map[string]string{
    utils.VolumeSnapshotContentManagedByLabel: nodeName,
  }
}
```
Since we know the side car is running on this node, and we know the snapshot content will reside on that particular node. I see no real reason to go through looking up the PV, then looking up all the nodes, and then matching the affinity of the PV to a node. When in the end we will just return NODE_NAME anyway.

awels replied:
Also are we sure there are no labels on the snapshotContent already? We are overriding anything that is there (if anything)

nearora-msft replied:
For your 1st question- the common snapshot_controller will still run as a single instance and not on every node. This controller is responsible for creating VolumeSnapshotContent, so it will figure out the right node and will label the VolumeSnapshotContent instance with that node.
The csi-snapshotter, on the other hand, will run as a daemon set on each node and will filter out the VolumeSnapshotContent objects by matching the node label.

For you 2nd question, we are adding a new label here, so is it possible that we could be overriding some other label?

awels replied:
Oh I totally missed this is the common controller, and not the side car. I figured I would optimize this and we wouldn't have to do all the lookups of the nodes and volumes. One more question on line 646 of the same file right before creating the snapshotcontent object this is called:
```go
class, volume, contentName, snapshotterSecretRef, err := ctrl.getCreateSnapshotInput(snapshot)
```
volume is the PV that we are looking up, could we pass the volume to the function so we can avoid having to look it up again? Secondly why are we looking up nodes to match names against. Regardless of the status of the node (down/up/gone) we would want to label the snapshot content with the node the original volume was on. So why not just look at the affinity of the PV, and return the node name from it. That way we don't have to look up the nodes.

For the second question, I just looked at the code above again, and the snapshotContent was just created a few lines above and it has no labels. So we are good.

nearora-msft replied:
Yeah, I suppose I could reuse the volume that we retrieve on line 646. I'll make the changes for that. Thanks for pointing it out.

For the next question, we are looking up nodes and then matching is against the node affinity and not simply getting the label from the volume and applying is to VolumeSnapshotContent is, because I don't think there's a global label key that all the drivers use to manage node affinity. Each driver will have its own topology keys, I suppose. So I wasn't sure what should be the label key which would give us the node name as value. Let me know if you can think of any better way to handle this.

awels replied:
You make a very good point about the topology key being different for each driver. For my driver I have been working on the reverse problem (restoring a snapshot into a volume). And I can make assumptions about the topology key in there and avoid the node lookup.

My main concern is the following (unlikely) scenario. The PV was created, we are about to look up the nodes, and the node we are interested went down, and is no longer in the list. So now when we loop over the nodes, we will never find the node that matches and thus we won't be able to label the snapshot content. And I would very much like to be able to count on the label being there, for the restore of a snapshot.

One problem I have not been able to solve is efficiently locate the node the snapshot is on if I want to restore it on another node (This can happen if the k8s scheduler schedules a pod that uses a WFFC PVC that has a datasource of a snapshot, on a different node than where the snapshot resides). If the label exists on the snapshot content, it would be trivial to locate the node.

I cannot think of a better mechanism than what you have done at this point, so I think I am happy right now with the code. I can just fail the restore if the node label is not there or empty.

awels commented:
/lgtm for what it is worth since I am not in the org.
@xing-yang what are we missing at this point?

xing-yang commented:
@zhucan Can you help test this?  It's rebased and updated now.

zhucan commented:
> @zhucan Can you help test this? It's rebased and updated now.

yes

zhucan commented on cmd/csi-snapshotter/main.go:
```diff
@@ -116,6 +127,19 @@ func main() {
 	coreFactory := coreinformers.NewSharedInformerFactory(kubeClient, *resyncPeriod)
+	var snapshotContentfactory informers.SharedInformerFactory
+	if *enableNodeDeployment {
+		node := os.Getenv("NODE_NAME")
```
It's KUBE_NODE_NAME, not NODE_NAME?

xing-yang replied:
I think this is a little confusing.
"KUBE_NODE_NAME" is used to populate the `nodeid` parameter for the CSI hostpath driver:
https://github.com/kubernetes-csi/csi-driver-host-path/blob/v1.7.3/cmd/hostpathplugin/main.go#L50
"NODE_NAME" is used for csi-snapshotter and it is consistent with what is used in external-provisioner:
https://github.com/kubernetes-csi/external-provisioner/blob/v3.0.0/cmd/csi-provisioner/csi-provisioner.go#L129
They should have the same value though, so we don't need two env variables.

It's fine to submit another PR to address this.  @zhucan Can you open an issue to rename "KUBE_NODE_NAME"  to "NODE_NAME" in setup-csi-snapshotter.yaml?  If you have time, please feel free to fix it.  Thanks.

zhucan replied:
@xing-yang ok

zhucan commented:
@nearora-msft I have other question, how to create a snapshot that the VolumeSnapshotContent has label with "snapshot.storage.kubernetes.io/managed-by=nodeName".

xing-yang commented:
> @nearora-msft I have other question, how to create a snapshot that the VolumeSnapshotContent has label with "snapshot.storage.kubernetes.io/managed-by=nodeName". And can you give the test yaml?

@zhucan I think the label is automatically set by snapshot-controller here during dynamic provisioning:
https://github.com/kubernetes-csi/external-snapshotter/pull/585/files#diff-1a3473f9016f7abcf5007414ff7141e39032aaf7aba079eeef262d04167208a2R681

zhucan commented:
@xing-yang I had tested on this pr.

Procedure：
**1. Deploy the external-provisioner with distributed mode.**
**2. Deploy hostpath driver with daemonset.**
**3. Deploy csi-snapshotter with distributed mode and deploy the snapshot-controller with deployment(replicas=2)**
**4. Create pvc, the responses of the pv is:(nodeAffinity is not nil)**
apiVersion: v1
kind: PersistentVolume
metadata:
  annotations:
    pv.kubernetes.io/provisioned-by: hostpath.csi.k8s.io
  creationTimestamp: "2021-11-16T03:41:22Z"
  finalizers:
  - kubernetes.io/pv-protection
  managedFields:
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .: {}
          f:pv.kubernetes.io/provisioned-by: {}
      f:spec:
        f:accessModes: {}
        f:capacity:
          .: {}
          f:storage: {}
        f:claimRef:
          .: {}
          f:apiVersion: {}
          f:kind: {}
          f:name: {}
          f:namespace: {}
          f:resourceVersion: {}
          f:uid: {}
        f:csi:
          .: {}
          f:driver: {}
          f:volumeAttributes:
            .: {}
            f:storage.kubernetes.io/csiProvisionerIdentity: {}
          f:volumeHandle: {}
        f:nodeAffinity:
          .: {}
          f:required:
            .: {}
            f:nodeSelectorTerms: {}
        f:persistentVolumeReclaimPolicy: {}
        f:storageClassName: {}
        f:volumeMode: {}
    manager: csi-provisioner
    operation: Update
    time: "2021-11-16T03:41:22Z"
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        f:phase: {}
    manager: kube-controller-manager
    operation: Update
    time: "2021-11-16T03:41:22Z"
  name: pvc-c5139089-9769-46ba-8be9-3c512c7abe6b
  resourceVersion: "29359541"
  selfLink: /api/v1/persistentvolumes/pvc-c5139089-9769-46ba-8be9-3c512c7abe6b
  uid: d7f18989-5619-421b-9050-a2d2e615ea0e
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  claimRef:
    apiVersion: v1
    kind: PersistentVolumeClaim
    name: csi-pvc
    namespace: default
    resourceVersion: "29359536"
    uid: c5139089-9769-46ba-8be9-3c512c7abe6b
  csi:
    driver: hostpath.csi.k8s.io
    volumeAttributes:
      storage.kubernetes.io/csiProvisionerIdentity: 1637032706867-8081-hostpath.csi.k8s.io-k8s-3.deeproute.ai
    volumeHandle: 8b741684-468d-11ec-bca8-fa1c22d69be7
  nodeAffinity:
    required:
      nodeSelectorTerms:
      - matchExpressions:
        - key: topology.hostpath.csi/node
          operator: In
          values:
          - **k8s-3.deeproute.ai**
  persistentVolumeReclaimPolicy: Delete
  storageClassName: csi-hostpath-sc
  volumeMode: Filesystem
status:
  phase: Bound

**5. create snapshot on the pvc, the responses of the content is:**
apiVersion: snapshot.storage.k8s.io/v1
kind: VolumeSnapshotContent
metadata:
  creationTimestamp: "2021-11-16T06:46:22Z"
  finalizers:
  - snapshot.storage.kubernetes.io/volumesnapshotcontent-bound-protection
  generation: 1
  labels:
    **snapshot.storage.kubernetes.io/managed-by: k8s-3.deeproute.ai**
  managedFields:
  - apiVersion: snapshot.storage.k8s.io/v1
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        .: {}
        f:creationTime: {}
        f:readyToUse: {}
        f:restoreSize: {}
        f:snapshotHandle: {}
    manager: csi-snapshotter
    operation: Update
    time: "2021-11-16T06:46:22Z"
  - apiVersion: snapshot.storage.k8s.io/v1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
          .: {}
          v:"snapshot.storage.kubernetes.io/volumesnapshotcontent-bound-protection": {}
        f:labels:
          .: {}
          f:snapshot.storage.kubernetes.io/managed-by: {}
      f:spec:
        .: {}
        f:deletionPolicy: {}
        f:driver: {}
        f:source:
          .: {}
          f:volumeHandle: {}
        f:volumeSnapshotClassName: {}
        f:volumeSnapshotRef:
          .: {}
          f:apiVersion: {}
          f:kind: {}
          f:name: {}
          f:namespace: {}
          f:resourceVersion: {}
          f:uid: {}
    manager: snapshot-controller
    operation: Update
    time: "2021-11-16T06:46:22Z"
  name: snapcontent-97e15540-4dcb-4325-b3f0-f832e851e521
  resourceVersion: "29403625"
  selfLink: /apis/snapshot.storage.k8s.io/v1/volumesnapshotcontents/snapcontent-97e15540-4dcb-4325-b3f0-f832e851e521
  uid: 7f5ecd75-98c2-452d-bf7c-ad0de3642876
spec:
  deletionPolicy: Delete
  driver: hostpath.csi.k8s.io
  source:
    volumeHandle: 8b741684-468d-11ec-bca8-fa1c22d69be7
  volumeSnapshotClassName: csi-hostpath-snapclass
  volumeSnapshotRef:
    apiVersion: snapshot.storage.k8s.io/v1
    kind: VolumeSnapshot
    name: new-snapshot-demo
    namespace: default
    resourceVersion: "29403615"
    uid: 97e15540-4dcb-4325-b3f0-f832e851e521
status:
  creationTime: 1637044529182996592
  readyToUse: true
  restoreSize: 1073741824
  snapshotHandle: 63f07a81-46a7-11ec-aa5c-6eec2f6ae768

The pvc was created on k8s-node3 in the data dir, like this:
![image](https://user-images.githubusercontent.com/13997135/141935552-8a986440-8bff-4ac7-b636-34b573b9e26c.png)

CC @xing-yang

zhucan commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -1655,3 +1666,27 @@ func (ctrl *csiSnapshotCommonController) checkAndSetInvalidSnapshotLabel(snapsho
+	}
+	nodeSelectorTerms := pv.Spec.NodeAffinity.Required
+
+	nodes, err := ctrl.client.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
```
I think it's no needed to list all nodes when creating volumesnapshotcontent everytimes.

nearora-msft replied:
We not only want to check whether that node exists or not, but more importantly,  we also need to find which node matches the affinity criteria. That's why we need to get the list of nodes to figure out which nodes match the criteria.

xing-yang replied:
@nearora-msft Can we do something similar to what is done in external-provisioner?
https://github.com/kubernetes-csi/external-provisioner/pull/524/files#diff-963c2f5e2076b500e70c47c13f135cdbb29b81f270bc76d1c99f823fecd1b510R271

nearora-msft replied:
@xing-yang , I'm not sure I completely understand your suggestion. One difference that I can see between csi-provisioner and snapshot-controller is that csi-provisioner is deployed on each node, so it can use the environment variable NODE_NAME to get the node details, while in this implementation, the snapshot-controller will run as a single central instance and doesn't have any information about NODE_NAME. 

Let me known in case you meant something else.

pohly replied:
If there's only a single instance of it, why does it need to do something different for distributed snapshotting? But I don't know the snapshotter code, so I might be missing something entirely.

awels replied:
So I am glad I am not the only one that got this wrong the first time, IMO @nearora-msft is right that we need to list all the nodes here. The controller code here is creating the VolumeSnapshotContent object, and then labeling it so that the side cars (csi driver or snapshot side car) can easily identify if they are the node needing to do something with it. 

In order to label the VolumeSnapshotContent with the correct node name, we look at the node affinity of the PV that is the source of the VolumeSnapshotContent. And match the affinity term there against nodes, if one of the nodes matches, we know that is the node we are looking for. 

Since each CSI driver can use its own term naming, we cant figure out ahead of time what that is, so we have to look for the node name, and if that matches we know we have the right term. So we have to fetch all the nodes, and compare until we find one that matches.

xing-yang replied:
@nearora-msft Can this be a node informer that is only created if distributed snapshotting is supported?

nearora-msft replied:
@xing-yang , getManagedByNode will only be invoked if distributed snapshotting is enabled. Is there a particular reason that we should create an informer instead of making the List call?

xing-yang replied:
@nearora-msft Informer should be used for better performance since this is invoked for every create snapshot call when distributed snapshotting is enabled.

nearora-msft replied:
Added an informer for listing nodes.

xing-yang commented:
@pohly Can you help review this PR?

xing-yang commented:
/assign @pohly

xing-yang commented:
@nearora-msft Can you add unit tests?

nearora-msft commented:
> @nearora-msft Can you add unit tests?

Added.

pohly commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -29,6 +29,7 @@ import (
 	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/client-go/kubernetes/scheme"
 	ref "k8s.io/client-go/tools/reference"
+	corev1 "k8s.io/component-helpers/scheduling/corev1"
```
`corev1` is usually used for the v1 core Kubernetes API. This package here gets imported as `corev1helpers` or `v1helpers` in k/k.

nearora-msft replied:
Changed to corev1helpers

xing-yang commented on pkg/common-controller/snapshot_controller.go:
```diff
@@ -671,6 +672,16 @@ func (ctrl *csiSnapshotCommonController) createSnapshotContent(snapshot *crdv1.V
 		},
 	}
 
+	nodeName, err := ctrl.getManagedByNode(volume)
```
Why do we need to get nodeName if distributed snapshotting is not supported? I think this is going have impact on deployment that does have support distributed snapshotting as well.

nearora-msft replied:
We could probably use a flag for snapshot controller to indicate whether we are using distributed snapshotting or not. But, since only a single implementation of snapshot controller will be deployed even in case of distributed snapshotting, I thought a flag like this might cause confusion.

 Let me know what you think. I could add a flag if it impacts the deployment that doesn't support distributed snapshotting.

xing-yang replied:
Yes, please add a flag to indicate distributed snapshotting is supported and set it to false by default.

nearora-msft replied:
Added the flag to indicate that distributed snapshotting is enabled.

xing-yang commented:
@nearora-msft Can you address the review comments?  I'd like to get this in the 5.0.0 release.  Thanks.

nearora-msft commented:
/retest

nearora-msft commented:
/retest

nearora-msft commented:
> @nearora-msft Can you address the review comments? I'd like to get this in the 5.0.0 release. Thanks.

Addressed the comments and tested with the latest revision.

xing-yang commented:
/test pull-kubernetes-csi-external-snapshotter-1-23-on-kubernetes-master

xing-yang commented:
/test pull-kubernetes-csi-external-snapshotter-alpha-1-22-on-kubernetes-1-22

xing-yang commented on pkg/utils/util.go:
```diff
@@ -106,7 +106,8 @@ const (
 	// See https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/177-volume-snapshot/tighten-validation-webhook-crd.md#automatic-labelling-of-invalid-objects
-	VolumeSnapshotInvalidLabel = "snapshot.storage.kubernetes.io/invalid-snapshot-resource"
+	VolumeSnapshotInvalidLabel          = "snapshot.storage.kubernetes.io/invalid-snapshot-resource"
+	VolumeSnapshotContentManagedByLabel = "snapshot.storage.kubernetes.io/managed-by"
```
Can you add a comment to explain what this label is for?

xing-yang replied:
Please add something like this to clarify what is in this label: `It specifies the node name which handles the snapshot for the volume local to that node.`

nearora-msft replied:
Done

pohly commented on cmd/csi-snapshotter/main.go:
```diff
@@ -75,11 +79,12 @@ var (
+	metricsPath          = flag.String("metrics-path", "/metrics", "The HTTP path where prometheus metrics will be exposed. Default is `/metrics`.")
+	retryIntervalStart   = flag.Duration("retry-interval-start", time.Second, "Initial retry interval of failed volume snapshot creation or deletion. It doubles with each failure, up to retry-interval-max. Default is 1 second.")
+	retryIntervalMax     = flag.Duration("retry-interval-max", 5*time.Minute, "Maximum retry interval of failed volume snapshot creation or deletion. Default is 5 minutes.")
+	enableNodeDeployment = flag.Bool("node-deployment", false, "Enables deploying the sidecar controller together with a CSI driver on nodes to manage snapshots for node-local volumes.")
```
Does it make sense to run the snapshotter with `--node-deployment` and without `--enable-distributed-snapshotting`?

If no, then `--node-deployment` can be removed and it can be enabled together with `--enable-distributed-snapshotting`.

xing-yang replied:
We need to have a flag to determine whether the sidecar is enabled for distributed snapshotting and it needs to be disabled by default. With a new feature like this, we can't enable it by default.

@nearora-msft, can you add a section in README with the title "Distributed Snapshotting", and clarify both flags are needed for this to work?

nearora-msft replied:
Added here : https://github.com/nearora-msft/external-snapshotter/blob/implement-distributed-snapshotting/README.md#distributed-snapshotting

xing-yang commented:
/test pull-kubernetes-csi-external-snapshotter-1-23-on-kubernetes-master

xing-yang commented:
/test pull-kubernetes-csi-external-snapshotter-alpha-1-22-on-kubernetes-1-22

xing-yang commented:
/lgtm
/approve

xing-yang commented:
@nearora-msft I'm merging the PR.  Thanks for your hard work!

Can you add some description for this `enable-distributed-snapshotting` flag in snapshot controller in your PR description?

Please also backport this PR to release-5.0 branch.

nearora-msft commented:
> @nearora-msft I'm merging the PR. Thanks for your hard work!
> 
> Can you add some description for this `enable-distributed-snapshotting` flag in snapshot controller in your PR description?
> 
> Please also backport this PR to release-5.0 branch.

Sure

awels commented:
Yes @nearora-msft thank you for the work. Have you thought about the reverse problem? Now that we can properly take snapshots, we need to restore them. In my testing with WFFC storage I found that we have a similar problem on restoring (same problem exists for csi clone). What happens if the scheduler schedules the pod that uses the new PVC that restores on a node that is not the same node as where the snapshot resides.

There are 2 options on solving it, but only one makes good sense IMO:
1. The CSI driver has some mechanism of copying data from other nodes to restore. I don't think CSI drivers want to be in the business of copying data from node to node.
2. Make the k8s scheduler aware of distributed snapshotting/csi clone, and implement a node filter based on the dataSource of the PVC and force the scheduler to only schedule pods on the node that contains the dataSource.

2 makes the most sense to me, but I am writing just in case I completely missed something obvious that solves the restore problem.

nearora-msft commented:
> 

@awels Yes, 2 makes more sense to me as well.

As per this implementation of [distributed provisioning](https://github.com/kubernetes-csi/external-provisioner/pull/524/files), for local volumes, if binding mode is set to WaitForFirstConsumer, the pvc will contain the name of the node in the selectedNode annotation and the pv will volume would be created on that node. I can see this annotation being set here(https://github.com/kubernetes/kubernetes/blob/f9e4cfcf51e3e0bbe0579c41e92c93aa2ed0c887/pkg/scheduler/framework/plugins/volumebinding/binder.go#L419). 

We could probably do something similar for snapshot restore. We could set the same annotation in case of snapshot restore, regardless of the binding mode. The annotation will contain info about the node that is handling the snapshot and the external-provisioner will make sure that the snapshot is restored on that node. We'd probably have to make changes in the external-provisioner as well.

@pohly please correct me if I didn't interpret this correctly.

awels commented:
> As per this implementation of [distributed provisioning](https://github.com/kubernetes-csi/external-provisioner/pull/524/files), for local volumes, if binding mode is set to WaitForFirstConsumer, the pvc will contain the name of the node in the selectedNode annotation and the pv will volume would be created on that node. I can see this annotation being set [here](https://github.com/kubernetes/kubernetes/blob/f9e4cfcf51e3e0bbe0579c41e92c93aa2ed0c887/pkg/scheduler/framework/plugins/volumebinding/binder.go#L419).

Yes, but AFAICT there is no mechanism to ensure that the nodeName being set by the scheduler is the same node as the dataSource source (PVC or snapshot). Which is where my filter comment came from. IMO we need to solve the problem in the scheduler, which will then put the right node in the annotation, and from there everything should work. Now making a scheduler extension that does this doesn't seem terribly hard, the problem is in getting all the pods being created to use the extension when a csi driver that needs it is in the cluster, that is the part I am struggling with.

nearora-msft commented:
> 

Ah, ok. Got it. I can't think of a solution at the top of my head but will update the thread if I can think of something.

nearora-msft commented:
> > As per this implementation of [distributed provisioning](https://github.com/kubernetes-csi/external-provisioner/pull/524/files), for local volumes, if binding mode is set to WaitForFirstConsumer, the pvc will contain the name of the node in the selectedNode annotation and the pv will volume would be created on that node. I can see this annotation being set [here](https://github.com/kubernetes/kubernetes/blob/f9e4cfcf51e3e0bbe0579c41e92c93aa2ed0c887/pkg/scheduler/framework/plugins/volumebinding/binder.go#L419).
> 
> Yes, but AFAICT there is no mechanism to ensure that the nodeName being set by the scheduler is the same node as the dataSource source (PVC or snapshot). Which is where my filter comment came from. IMO we need to solve the problem in the scheduler, which will then put the right node in the annotation, and from there everything should work. Now making a scheduler extension that does this doesn't seem terribly hard, the problem is in getting all the pods being created to use the extension when a csi driver that needs it is in the cluster, that is the part I am struggling with.

Ah, ok. Got it. I can't think of a solution at the top of my head but will update the thread if I can think of something.
